File: _combination_sum.js
function combinationSum(candidates: number[], target: number): number[][] {

    let res = []

    let dfs = (i,cur,total)=>{

        if( total== target ){
            res.push(Array.from(cur))
            return
        }
        if(total>target || i>=candidates.length)
            return
        
        cur.push(candidates[i])
        dfs(i, cur, total+candidates[i])
        cur.pop()
        dfs(i+1, cur, total) 

    }

    dfs(0,[],0)

    return res

};

File: Python/combinationSum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        global res
        res = []

        def util(i, cur, total):
            if(i>=len(candidates) or total>target):
                return 
            
            if(total == target):
                res.append(cur.copy())
                return
            
            cur.append(candidates[i])
            util(i, cur, total + candidates[i])
            cur.pop()
            util(i+1, cur, total)
        
        util(0,[],0)
        return res
